### PR TITLE
Dtr 5260

### PR DESCRIPTION
**Pull request by Mounkumar**

### CODE REVIEW

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 1]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      out.subcontractorCanBeDeleted mustBe true
+
+      verify(cs).close()
+    }
+
+    "returns false when flag is null" in {
+      val db   = mock[Database]
+      val conn = mock[Connection]
+      val cs   = mock[CallableStatement]
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
val rs3  = mock[ResultSet]
val rs4  = mock[ResultSet]
val rs5  = mock[ResultSet]
```

File: app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala [hunk 2]
```diff
@@ -1737,4 +1739,40 @@ class CisFormpRepository @Inject() (@NamedDatabase("cis") db: Database)(implicit
...
+          cs.setString(1, cisId)
+          cs.setLong(2, subbieResourceRef)
+
+          cs.registerOutParameter(3, OracleTypes.CURSOR)
+          cs.registerOutParameter(4, OracleTypes.CURSOR)
+          cs.registerOutParameter(5, OracleTypes.CURSOR)
+          cs.registerOutParameter(6, Types.VARCHAR)
+
+          cs.execute()
+
```
Comment (jassalrichy) [inline app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala:1762]: Add the following after `cs.execute `

```
discardCursor(cs, 3)
discardCursor(cs, 4)
discardCursor(cs, 5)
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 3]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      when(db.withConnection(anyArg[Connection => Any])).thenAnswer { inv =>
+        val f = inv.getArgument(0, classOf[Connection => Any])
+        f(conn)
+      }
+
+      val call = "{ call SUBCONTRACTOR_PROCS.Get_Subcontractor_For_Delete(?, ?, ?, ?, ?, ?) }"
+      when(conn.prepareCall(eqTo(call))).thenReturn(cs)
+
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala:4064]: Add:

```
when(cs.getObject(3, classOf[ResultSet])).thenReturn(rs3)                                                                                                                                     
when(cs.getObject(4, classOf[ResultSet])).thenReturn(rs4)                                                                                                                                     
when(cs.getObject(5, classOf[ResultSet])).thenReturn(rs5) 
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 4]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
       verify(conn, never).prepareCall(eqTo(CisStoredProcedures.CallUpdateSubmission))
     }
   }
+
+  "getSubcontractorForDelete" - {
+
+    "calls stored procedure and returns true when flag is 'true'" in {
+      val db   = mock[Database]
+      val conn = mock[Connection]
+      val cs   = mock[CallableStatement]
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
val rs3  = mock[ResultSet]
val rs4  = mock[ResultSet]
val rs5  = mock[ResultSet]
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 5]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      when(db.withConnection(anyArg[Connection => Any])).thenAnswer { inv =>
+        val f = inv.getArgument(0, classOf[Connection => Any])
+        f(conn)
+      }
+
+      val call = "{ call SUBCONTRACTOR_PROCS.Get_Subcontractor_For_Delete(?, ?, ?, ?, ?, ?) }"
+      when(conn.prepareCall(eqTo(call))).thenReturn(cs)
+
+      when(cs.getString(6)).thenReturn("true")
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
when(cs.getObject(3, classOf[ResultSet])).thenReturn(rs3)
when(cs.getObject(4, classOf[ResultSet])).thenReturn(rs4)
when(cs.getObject(5, classOf[ResultSet])).thenReturn(rs5)
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 6]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+      verify(cs).registerOutParameter(6, Types.VARCHAR)
+
+      verify(cs).execute()
+      verify(cs).close()
+    }
+
+    "returns false when flag is 'false'" in {
+      val db   = mock[Database]
+      val conn = mock[Connection]
+      val cs   = mock[CallableStatement]
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
val rs3  = mock[ResultSet]
val rs4  = mock[ResultSet]
val rs5  = mock[ResultSet]
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 7]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      when(db.withConnection(anyArg[Connection => Any])).thenAnswer { inv =>
+        val f = inv.getArgument(0, classOf[Connection => Any])
+        f(conn)
+      }
+
+      val call = "{ call SUBCONTRACTOR_PROCS.Get_Subcontractor_For_Delete(?, ?, ?, ?, ?, ?) }"
+      when(conn.prepareCall(eqTo(call))).thenReturn(cs)
+
+      when(cs.getString(6)).thenReturn("false")
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
when(cs.getObject(3, classOf[ResultSet])).thenReturn(rs3)
when(cs.getObject(4, classOf[ResultSet])).thenReturn(rs4)
when(cs.getObject(5, classOf[ResultSet])).thenReturn(rs5)
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 8]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      val repo = new CisFormpRepository(db)
+
+      val out = repo.getSubcontractorForDelete("cis-123", 10L).futureValue
+
+      out.subcontractorCanBeDeleted mustBe false
+
+      verify(cs).execute()
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Update to:

```
verify(cs).execute()
verify(cs).getObject(3, classOf[ResultSet])
verify(cs).getObject(4, classOf[ResultSet])
verify(cs).getObject(5, classOf[ResultSet])
verify(cs).close()
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 9]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+      verify(cs).close()
+    }
+
+    "maps trimmed 'true' as true" in {
+      val db   = mock[Database]
+      val conn = mock[Connection]
+      val cs   = mock[CallableStatement]
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
val rs3  = mock[ResultSet]
val rs4  = mock[ResultSet]
val rs5  = mock[ResultSet]
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala [hunk 10]
```diff
@@ -3919,4 +3919,115 @@ final class CisFormpRepositorySpec extends SpecBase {
...
+
+      when(db.withConnection(anyArg[Connection => Any])).thenAnswer { inv =>
+        val f = inv.getArgument(0, classOf[Connection => Any])
+        f(conn)
+      }
+
+      val call = "{ call SUBCONTRACTOR_PROCS.Get_Subcontractor_For_Delete(?, ?, ?, ?, ?, ?) }"
+      when(conn.prepareCall(eqTo(call))).thenReturn(cs)
+
+      when(cs.getString(6)).thenReturn(" true ")
```
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala]: Add:

```
when(cs.getObject(3, classOf[ResultSet])).thenReturn(rs3)
when(cs.getObject(4, classOf[ResultSet])).thenReturn(rs4)
when(cs.getObject(5, classOf[ResultSet])).thenReturn(rs5)
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala:1]: CisFormpRepository added parameters 3, 4, and 5 are registered as OracleTypes.CURSOR but after cs.execute() they are never touched. The rest of the codebase has a discardCursor helper for exactly this pattern. When a stored procedure writes a REF CURSOR into an OUT parameter, Oracle allocates a server-side cursor for it. Calling cs.close() via Using.resource in withCall closes the statement but does not guarantee the individual REF CURSOR resources are released - discardCursor does cs.getObject(index, classOf[ResultSet]) and then Using.resource(rs)(_ => ()) to explicitly close each one.

File: app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala [hunk 11]
```diff
@@ -1737,4 +1739,40 @@ class CisFormpRepository @Inject() (@NamedDatabase("cis") db: Database)(implicit
...
+          val subcontractorCanBeDeleted =
+            Option(cs.getString(6)).map(_.trim.toLowerCase) match {
+              case Some("true")  => true
+              case Some("false") => false
+              case _             => false
```
Comment (jassalrichy) [inline app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala]: Update wildcard case:

from:
`case _             => false`

to:
```
case other =>                                                                                                                                                                                                 
logger.warn(s"[getSubcontractorForDelete] unexpected flag value: '$other'")                                                                                                                                 
false    
```

File: test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepositorySpec.scala:1]: Add the following test:

```
  "throws IllegalStateException when multiple subcontractors are returned" in new Ctx {                                                                                                     
                                                                                                                                                                                            
    when(rs4.next()).thenReturn(true, true, false)                                                                                                                                          
    // minimal column stubs for two rows                                                                                                                                                    
    when(rs4.getLong("subcontractor_id")).thenReturn(1L, 2L)                                                                                                                                
    when(rs4.getLong("subbie_resource_ref")).thenReturn(10L, 11L)                                                                                                                           
    when(rs4.getString("type")).thenReturn("company")                                                                                                                                       
    when(rs4.getString("tradingname")).thenReturn("Gamma Builders")                                                                                                                         
    when(rs4.wasNull()).thenReturn(true)                                                                                                                                                    
    Seq("utr", "partner_utr", "crn", "firstname", "nino", "secondname", "surname",                                                                                                          
        "partnership_tradingname", "address_line_1", "address_line_2", "address_line_3",                                                                                                    
        "address_line_4", "country", "postcode", "email_address", "phone_number",                                                                                                           
        "mobile_phone_number", "works_reference_number", "tax_treatment",                                                                                                                   
        "updated_tax_treatment", "verification_number", "matched", "verified",                                                                                                              
        "auto_verified").foreach(f => when(rs4.getString(f)).thenReturn(null))                                                                                                              
    when(cs.getString(6)).thenReturn("true")                                                                                                                                                
                                                                                                                                                                                            
    val exception =                                                                                                                                                                         
      repo.getSubcontractorForDelete("cis-123", 10L).failed.futureValue                                                                                                                     
                                                                                                                                                                                            
    exception mustBe an[IllegalStateException]                                                                                                                                              
    exception.getMessage must include("Expected exactly one subcontractor")                                                                                                                 
  }    
```

File: app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala [hunk 12]
```diff
@@ -1737,4 +1739,58 @@ class CisFormpRepository @Inject() (@NamedDatabase("cis") db: Database)(implicit
...
+          cs.registerOutParameter(5, OracleTypes.CURSOR)
+          cs.registerOutParameter(6, Types.VARCHAR)
+
+          cs.execute()
+
+          discardCursor(cs, 3)
+          discardCursor(cs, 5)
+
+          val subcontractor             =
+            withCursor(cs, 4)(collectSubcontractors) match {
```
Comment (jassalrichy) [inline app/uk/gov/hmrc/formpproxy/cis/repositories/CisFormpRepository.scala:1767]: Non-exhaustive match on cursor 4 results. If the stored procedure ever returns 2+ rows this throws a MatchError.

Add a third case for multiple results:                                                                                                                                                             
```
case multiple =>                                                                                                                                                  
  throw new IllegalStateException(                                                                                                                                
  s"Expected exactly one subcontractor for cisId=$cisId and subbieResourceRef=$subbieResourceRef, got ${multiple.size}"                                         
  )   
```

File: test/uk/gov/hmrc/formpproxy/cis/models/SubcontractorSpec.scala
Comment (jassalrichy) [inline test/uk/gov/hmrc/formpproxy/cis/models/SubcontractorSpec.scala:1]: 'Missing displayName edge case test: No test covers a SoleTrader with all name fields empty and no tradingName - it should fall through to "No name provided" but isn't asserted.'

Add the following after line 180:

```
"return default text for a sole trader with no name fields and no trading name" in {                                                                                         
        val subcontractor =                                                                                                                                                        
        baseSubcontractor.copy(                                                                                                                                                  
        subcontractorType = Some("soletrader"),                                                                                                                                
        firstName = None,                                                                                                                                                      
        surname = None,                                                                                                                                                        
        tradingName = None                                                                                                                                                     
        )                                                                                                                                                                        
        subcontractor.displayName mustBe "No name provided"                                                                                                                        
}    
                                                                                                                                                                        
 "return default text when subcontractor type is unrecognised" in {                                                                                                           
        val subcontractor =                                                                                                                                                        
        baseSubcontractor.copy(                                                                                                                                                  
        subcontractorType = Some("unknown"),                                                                                                                                   
        firstName = Some("Alan"),                                                                                                                                              
        surname = Some("Grant"),                                                                                                                                               
        tradingName = Some("Grant Builders")                                                                                                                                   
        )                                                                                                                                                                        
        subcontractor.displayName mustBe "No name provided"                                                                                                                        
}    
```
